internal/grep/grep.go

```diff
@@ -0,0 +1,274 @@
+package grep
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// Result represents the result of a grep operation
+type Result struct {
+	ExitCode int
+	Stdout   []byte
+	Stderr   []byte
+}
+
+// EmulateGrep provides a simplified interface that mimics grep command behavior
+func EmulateGrep(args []string, stdin []byte) Result {
+	if len(args) == 0 {
+		return Result{
+			ExitCode: 2,
+			Stderr:   []byte("Usage: grep [options] pattern [files...]\n"),
+		}
+	}
+
+	// Parse arguments
+	var recursive bool
+	var pattern string
+	var files []string
+	var useStdin = true
+
+	for i := 0; i < len(args); i++ {
+		arg := args[i]
+		switch arg {
+		case "-r":
+			recursive = true
+		case "-E":
+			// Extended regex flag - ignore for now as we always use extended
+			if i+1 < len(args) {
+				i++
+				pattern = args[i]
+			}
+		default:
+			if pattern == "" {
+				pattern = arg
+			} else {
+				files = append(files, arg)
+				useStdin = false
+			}
+		}
+	}
+
+	if pattern == "" {
+		return Result{
+			ExitCode: 2,
+			Stderr:   []byte("No pattern specified\n"),
+		}
+	}
+
+	if useStdin {
+		return searchStdin(pattern, string(stdin))
+	} else {
+		return searchFiles(pattern, files, options{recursive: recursive})
+	}
+}
+
+type options struct {
+	recursive bool
+}
+
+type matcher interface {
```

Comment:
This interface looks like pointless indirection - just use the `matchWithBackreferences` function directly? Looks like `match` just delegates to that anyway.